src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonIEnumerableConverterFactory.cs

```diff
@@ -0,0 +1,234 @@
+                elementType = actualTypeToConvert.GetGenericArguments()[0];
+            }
+            // ConcurrentStack<> or deriving from ConcurrentStack<>
+            else if ((actualTypeToConvert = typeToConvert.GetCompatibleGenericBaseClass(typeof(ConcurrentStack<>))) != null)
+            {
+                converterType = typeof(JsonConcurrentStackOfTConverter<,>);
+                elementType = actualTypeToConvert.GetGenericArguments()[0];
+            }
+            // ConcurrentQueue<> or deriving from ConcurrentQueue<>
+            else if ((actualTypeToConvert = typeToConvert.GetCompatibleGenericBaseClass(typeof(ConcurrentQueue<>))) != null)
+            {
+                converterType = typeof(JsonConcurrentQueueOfTConverter<,>);
+                elementType = actualTypeToConvert.GetGenericArguments()[0];
+            }
+            // IEnumerable<>, types assignable from List<>
+            else if ((actualTypeToConvert = typeToConvert.GetCompatibleGenericInterface(typeof(IEnumerable<>))) != null)
+            {
+                converterType = typeof(JsonIEnumerableOfTConverter<,>);
+                elementType = actualTypeToConvert.GetGenericArguments()[0];
+            }
+            // Check for non-generics after checking for generics.
+            else if (typeof(IDictionary).IsAssignableFrom(typeToConvert))
+            {
+                if (typeToConvert == typeof(IDictionary))
+                {
+                    return s_IDictionaryConverter;
+                }
+
+                converterType = typeof(JsonIDictionaryConverter<>);
+            }
+            else if (typeof(IList).IsAssignableFrom(typeToConvert))
+            {
+                if (typeToConvert == typeof(IList))
+                {
+                    return s_IListConverter;
+                }
+
+                converterType = typeof(JsonIListConverter<>);
+            }
+            else if (typeToConvert.IsNonGenericStackOrQueue())
+            {
+                converterType = typeof(JsonIEnumerableWithAddMethodConverter<>);
+            }
+            else
+            {
+                Debug.Assert(typeof(IEnumerable).IsAssignableFrom(typeToConvert));
+                if (typeToConvert == typeof(IEnumerable))
+                {
+                    return s_IEnumerableConverter;
+                }
+
+                converterType = typeof(JsonIEnumerableConverter<>);
+            }
+
+            if (converterType != null)
+            {
+                Type genericType;
+                if (converterType.GetGenericArguments().Length == 1)
+                {
+                    genericType = converterType.MakeGenericType(typeToConvert);
+                }
+                else
+                {
+                    genericType = converterType.MakeGenericType(typeToConvert, elementType!);
+                }
+
+                converter = (JsonConverter)Activator.CreateInstance(
```

Comment:
Actually, here is a special case, where `Activator.CreateInstance` returns nullable, because you can call `Activator.CreateInstance(typeof(Nullable<int>)...)` and it could return null in those cases. However, in this case we know that `Activator.CreateInstance` will not return null, so I think it makes the most sense to use `!` for that and cast to non-nullable. So either way is fine, but I think expressing it this way, tells developers, if we get through this path, converter will not be null.

Reply:
Good point. `typeToConvert` can be `Nullable<T>` but that is only the generic argument of `converterType` in this case which itself is some `IEnumerable<T>`). So, we will never have `Activator.CreateInstance(typeof(Nullable<int>)...)` in this case (but might have `Activator.CreateInstance(typeof(List<Nullable<int>>)...)`, which is fine).

Thanks @safern